astropy/units/utils.py

```diff
@@ -293,3 +294,166 @@ def quantity_asanyarray(a, dtype=None):
         return Quantity(a, dtype=dtype)
     else:
         return np.asanyarray(a, dtype=dtype)
+
+
+# ------------------------------------------------------------------------------
+
+
+def frompyfunc(func, nin, nout, inunits=None, ounits=None,
+               *, identity=None, assume_correct_units=False):
+    """Quantity-aware `numpy.frompyfunc`. Works as a drop-in replacement.
+
+    `~numpy.ufunc`s operate on only recognized `~numpy.dtype`s (e.g. float32),
+    so units must be removed beforehand and replaced afterwards. Therefore units
+    MUST BE KNOWN a priori, as they will not be propagated by the astropy
+    machinery.
+
+    Parameters
+    ----------
+    func : callable
+    nin, nout : int
+        Number of ufunc's inputs and outputs.
+    inunits, ounits : unit-like or sequence thereof or None (optional)
+        Sequence of the input and output units, respectively.
+
+        .. warning::
+            Inputs will be converted to these units before being passed to the
+            returned `~numpy.ufunc`. Outputs will be assigned these units.
+            Make sure these are the correct units.
+
+    identity : object (optional, keyword-only)
+        The value to use for the `~numpy.ufunc.identity` attribute of the
+        resulting object. If specified, this is equivalent to setting the
+        underlying C ``identity`` field to ``PyUFunc_IdentityValue``.
+        If omitted, the identity is set to ``PyUFunc_None``. Note that this is
+        _not_ equivalent to setting the identity to ``None``, which implies the
+        operation is reorderable.
+
+    assume_correct_units : bool (optional, keyword-only)
```

Comment:
Not quite sure about this. I feel that if it is needed, people could just write their own helper. If we do keep it, maybe have it be a more generic `default_units`? Could be `True` for `inunits`, a list of not equal, `u.one` for dimensionless (default) and `False` for not allowed.